app/src/main/java/de/westnordost/streetcomplete/data/user/oauth/OAuthAuthorization.kt

```diff
@@ -0,0 +1,203 @@
+package de.westnordost.streetcomplete.data.user.oauth
+
+import de.westnordost.streetcomplete.ApplicationConstants
+import de.westnordost.streetcomplete.data.download.ConnectionException
+import de.westnordost.streetcomplete.data.user.AuthorizationException
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.Transient
+import kotlinx.serialization.json.Json
+import java.net.HttpURLConnection
+import java.net.URI
+import java.net.URL
+import java.net.URLDecoder
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import kotlin.io.encoding.Base64
+import kotlin.io.encoding.ExperimentalEncodingApi
+
+/** One authorization with OAuth. For each authorization request, a new instance must be created
+ *
+ *  Authorization flow:
+ *
+ *  1. createAuthorizationUrl() and open it in a browser (or web view)
+ *  2. let user accept or deny the permission request in the browser. Authorization endpoint will
+ *     call the redirect URI (aka callback URI) with parameters in either case.
+ *  3. check if the URI received is for this instance with itsForMe(uri) and then
+ *     extractAuthorizationCode(uri) from that URI
+ *  4. retrieveAccessToken(authorizationCode) with the retrieved authorizationCode
+ *
+ *
+ *  @param authorizationUrl the OAuth2 server authorization endpoint
+ *  @param accessTokenUrl the OAuth2 server token endpoint
+ *  @param clientId the OAuth2 client id
+ *  @param scopes the scopes (aka permissions) to request
+ *  @param redirectUri the redirect URI (aka callback URI) that the authorization endpoint should
+ *                     call when the user allowed the authorization.
+ * */
```

Comment:
```suggestion
/**
 * One authorization with OAuth. For each authorization request, a new instance must be created
 *
 * Authorization flow:
 *
 * 1. createAuthorizationUrl() and open it in a browser (or web view)
 * 2. let user accept or deny the permission request in the browser. Authorization endpoint will
 *    call the redirect URI (aka callback URI) with parameters in either case.
 * 3. check if the URI received is for this instance with itsForMe(uri) and then
 *    extractAuthorizationCode(uri) from that URI
 * 4. retrieveAccessToken(authorizationCode) with the retrieved authorizationCode
 *
 *
 * @param authorizationUrl the OAuth2 server authorization endpoint
 * @param accessTokenUrl the OAuth2 server token endpoint
 * @param clientId the OAuth2 client id
 * @param scopes the scopes (aka permissions) to request
 * @param redirectUri the redirect URI (aka callback URI) that the authorization endpoint should
 *                    call when the user allowed the authorization.
 */
```

see https://kotlinlang.org/docs/kotlin-doc.html